2.Para el siguiente conjunto de datos, obtener la Matriz HAT, H, los valores predichos, y ̂, y los coeficientes de la ecuación de regresión β.

X1: 2 3 5 7 9

y: 5 8 7 10 12

In [9]:
import numpy as np

X = np.array([[1,2],
              [1,3],
              [1,5],
              [1,7],
              [1,9]
              ])

y = np.array([5, 8, 7, 10, 12])

# HAT matrix calulation
H = X @ np.linalg.inv(X.T @ X) @ X.T
print("Matriz Hat (H):")
print(H)

# β_hat (coeficientes de regresión)
X_transpose = X.T
beta_hat = np.linalg.inv(X_transpose @ X) @ X_transpose @ y

print("\nCoeficientes de regresión (β_hat):")
print(beta_hat)

# y_hat (valores predichos)
y_hat = X @ beta_hat

print("\nValores predichos (y_hat):")
print(y_hat)


Matriz Hat (H):
[[ 0.51219512  0.41463415  0.2195122   0.02439024 -0.17073171]
 [ 0.41463415  0.34756098  0.21341463  0.07926829 -0.05487805]
 [ 0.2195122   0.21341463  0.20121951  0.18902439  0.17682927]
 [ 0.02439024  0.07926829  0.18902439  0.29878049  0.40853659]
 [-0.17073171 -0.05487805  0.17682927  0.40853659  0.6402439 ]]

Coeficientes de regresión (β_hat):
[3.86585366 0.87195122]

Valores predichos (y_hat):
[ 5.6097561   6.48170732  8.22560976  9.9695122  11.71341463]


3.Utilizando los Datos “Cirugía de Hígado” obtener la matriz HAT, los valores predichos, y ̂, y los coeficientes de la ecuación de regresión β utilizando el método de matrices. (Puede realizarse con cualquier paquete).

Estandarizar las variables con MinMax Scaler antes de realizar las operaciones con matrices.

In [2]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression

In [3]:
df = pd.read_excel('sample_data/HAT.xlsx')

In [4]:
from sklearn.preprocessing import MinMaxScaler

X = df.iloc[:, :-1]
y = df.iloc[:, -1]

# Estandariza las características
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

In [5]:
X_b = np.hstack([np.ones((X_scaled.shape[0], 1)), X_scaled])  # Agrega columna de unos

In [8]:
#matriz HAT
#np.set_printoptions(threshold=np.inf)
np.set_printoptions(threshold=1000)

H = X_b @ np.linalg.inv(X_b.T @ X_b) @ X_b.T
print("Matriz HAT:")
print(H)

Matriz HAT:
[[ 0.03714196  0.01278513  0.03153001 ... -0.0107616   0.03002929
   0.00244981]
 [ 0.01278513  0.06106075  0.04827415 ...  0.02342793  0.0143413
   0.03831749]
 [ 0.03153001  0.04827415  0.09511866 ...  0.02121118  0.02055136
   0.03079561]
 ...
 [-0.0107616   0.02342793  0.02121118 ...  0.10349403  0.00417997
   0.03037262]
 [ 0.03002929  0.0143413   0.02055136 ...  0.00417997  0.06246334
  -0.00802438]
 [ 0.00244981  0.03831749  0.03079561 ...  0.03037262 -0.00802438
   0.05864247]]


In [ ]:
# Calcular los valores predichos
y_hat = H @ y.values
print("Valores predichos (y_hat):")
print(y_hat)

Valores predichos (y_hat):
[ 706.25623722  430.82292124  732.22983195  425.03957611 1454.587552
  317.98512422  561.43434506  620.57029786  814.99238167  761.68391557
  989.9519364   228.02181205 1394.13461057 1050.29172024  937.95668893
  732.48434715  418.09976441   58.32293366  857.99665125  939.03472439
  555.53153245  366.02513011  485.52130235  817.77817446  861.31877768
  669.35808718  592.75876392 1619.88673751  399.48541635  596.03964243
  257.21831085  163.93595245  594.67195134 1143.59947369  512.12738287
  702.86128551  629.09497899  563.49249113  556.62971999  607.66774824
  423.46168884  569.50576666 1363.25897377  625.40911054  688.74643782
  590.69719298  948.046226   1333.78069235  712.00261732 1065.38071841
  462.70846831  649.64082214  690.30288166  958.05440017  330.98090005
  811.81615789  635.24982009  367.64186795  629.61451147  469.62155034
  154.73023565  369.81392347  403.26975861  902.08295831  600.59893803
  772.29123143  207.7278467   396.43186123 1229.8395

In [ ]:
# Calcular los coeficientes de regresión
beta = np.linalg.inv(X_b.T @ X_b) @ X_b.T @ y.values
print("Coeficientes de regresión (β):")
print(beta)

Coeficientes de regresión (β):
[-575.86374747  453.26085291  738.23244281  854.14304366  429.43164879
   25.63892773   13.09258075  -41.26764482  195.70703222]


4.Concluir sobre el significado de los valores de apalancamiento y formular la ecuación de regresión.

Los valores de apalancamiento se encuentran en la diagonal de la matriz H y nos indican cuanto influyen los puntos de datos en el modelo. Podemos observar en la matriz del higado que la mayoria de los valores son muy pequeños, menores a 0.1, esto es bueno ya que significa que todos los valores tienen una influencia normal sobre el modelo, si el valor fuera mas grande, cercano a 1, indicaria una influencia muy grande y puede indicar datos atipicos.

La ecuación de regresión seria la siguente:

y_hat = -575.86 + 453.26x1 + 738.23x2 + 854.14x3 + 429.43x4 +25.64x5 + 13.09x6 - 41.27x7 + 195.71x8

-575.86 es el valor predicho de y_hat cuando todas las variables independientes son iguales a cero.
Cada aumento en una variable representa un aumento de y_hat del tamaño del coeficiente de dicha variable, entre mas grande el coeficiente mas infuyente es esa variable.


